### PR TITLE
[IOTDB-2803] Implement create timeseries metadata operation through consensus layer in MPP framework

### PR DESCRIPTION
**Pull request by yifuzhou**
## Description
This pr aims to streamline the process of metadata operation through consensus layer, and take create timeseries as an example.
The process divided into two parts:
first, transfer user's input sql to PlanNode and package into framentInstance(https://github.com/apache/iotdb/pull/5344);
second, for the remote node, after receiving the fragmentInstance, deserialize it and run in the consensus layer and execute the operation from fragmentInstance, return execution result status response.

TODO plan:
1. I just transfer plannode to physcialplan to execute no-query operation. Later will directly use plannode as a parameter to execute no-query operation.
2. Use visitor pattern to implement schema write operation.

Further work:
After MPP's scheduler has finished and TFragmentInstance has been determined, I will use this thrift structure instread of custom serialization and deserialization.

### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 1]
```diff
@@ -31,6 +39,20 @@
...
+  private PlanExecutor executor;
+
+  public SchemaRegionStateMachine() {
+    try {
+      executor = new PlanExecutor();
+    } catch (QueryProcessException e) {
+      if (e.getErrorCode() != TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode()) {
+        logger.debug("Meet error while processing non-query.", e);
+      } else {
+        logger.warn("Meet error while processing non-query. ", e);
```
[the remark this answers is not in the document]
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:51]: Original coordinator execution logic, just keep it.

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 2]
```diff
@@ -47,7 +66,28 @@ public void stop() {}
...
+    return write(physicalPlan);
+  }
+
+  protected TSStatus write(PhysicalPlan plan) {
+    boolean isSuccessful;
+    try {
+      isSuccessful = executor.processNonQuery(plan);
+    } catch (QueryProcessException e) {
+      logger.error("{}: query process Error: ", IoTDBConstant.GLOBAL_DB_NAME, e);
+      return RpcUtils.getStatus(e.getErrorCode(), e.getMessage());
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:83]: Should we invoke the interfaces provided by SchemaRegion directly?

Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java]: fix.

File: server/src/main/java/org/apache/iotdb/db/metadata/Executor/NoQueryExecutorProxy.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * under the License.
+ */
+
+package org.apache.iotdb.db.metadata.Executor;
+
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.db.metadata.schemaregion.SchemaRegion;
+
+public class NoQueryExecutorProxy implements NoQueryExecutor {
+  private NoQueryExecutor noQueryExecutor;
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/metadata/Executor/NoQueryExecutorProxy.java:26]: Please rename this to ```SchemaQueryExecutorProxy``` and add some docs and comments to explain why we add this static proxy.

File: server/src/main/java/org/apache/iotdb/db/metadata/Executor/NoQueryExecutor.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.apache.iotdb.db.metadata.Executor;
+
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.db.metadata.schemaregion.SchemaRegion;
+
+public interface NoQueryExecutor {
+  void executor(SchemaRegion schemaRegion) throws MetadataException;
+}
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply server/src/main/java/org/apache/iotdb/db/metadata/Executor/NoQueryExecutor.java:27]: We can implement this as a common framework of schema query process. So please consider how to integrate the queries with return values. Do we need any genericity skills to support interfaces like ```<T> getResult()```.

**Review by MarcosZyk — COMMENTED** [review]
Wonderful idea for schema query execution. Here are some comments and suggestions. Please take a look.

File: server/pom.xml [hunk 5]
```diff
@@ -262,6 +262,11 @@
             <version>2.6</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>com.codahale.metrics</groupId>
+            <artifactId>metrics-core</artifactId>
```
Comment (wangchao316) [inline server/pom.xml:267]: @OneSizeFitsQuorum see this.  ratis need this jar?

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java [hunk 6]
```diff
@@ -59,6 +59,7 @@ public ByteString getContent() {
             // TODO Pooling
             ByteBuffer byteBuffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE);
             actualRequest.serializeRequest(byteBuffer);
+            byteBuffer.flip();
```
[the remark this answers is not in the document]
Comment (yifuzhou) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java:62]: ByteString.copyFrom need to read byteBuffer and copy it. must flip it.

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/AlterTimeSeriesNode.java [hunk 7]
```diff
@@ -238,6 +238,43 @@ public static AlterTimeSeriesNode deserialize(ByteBuffer byteBuffer) {
         new PlanNodeId(id), path, alterType, alterMap, alias, tagsMap, attributesMap);
   }
 
+  //  @Override
```
[the remark this answers is not in the document]
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/AlterTimeSeriesNode.java:241]: will fix in next pr, after streamline the process, will add the rest operation.

File: server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java [hunk 8]
```diff
@@ -43,39 +45,73 @@
...
+      LOGGER.error(msg);
+      response.setAccepted(false);
+      response.setMessage(msg);
+      return response;
+    }
+    consensusManager.addConsensusGroup(fragmentInstance.getRegionReplicaSet());
+
     switch (type) {
       case READ:
         ConsensusReadResponse readResp = ConsensusImpl.getInstance().read(groupId, request);
```
[the remark this answers is not in the document]
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java:92]: I would recommend to use consensusManager to package it. Maybe consider it after read operation implement.

File: server/pom.xml (hunk 5 above)
Comment (yifuzhou) [reply server/pom.xml:267]: In my computer, when I run UT test related to ratis consensus protocol, it tells me I miss this jar.

**Review by xingtanzjr — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]

**Comment by xingtanzjr** [comment]
As our talk before, the PlanNode which is used to execute WRITE operation need to implement a method `splitByPartition()`. We can merge this PR first and do the work in next PR

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java [hunk 9]
```diff
@@ -59,6 +59,7 @@ public ByteString getContent() {
...
             serializedContent = ByteString.copyFrom(byteBuffer);
             byteBuffer.flip();
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java:64]: It seems that we can remove this line?

File: server/pom.xml (hunk 5 above)
Comment (OneSizeFitsQuorum) [reply server/pom.xml:267]: seems weird

File: confignode/pom.xml [hunk 10]
```diff
@@ -75,6 +75,11 @@
             <groupId>commons-io</groupId>
             <artifactId>commons-io</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.codahale.metrics</groupId>
+            <artifactId>metrics-core</artifactId>
```
Comment (OneSizeFitsQuorum) [inline confignode/pom.xml:80]: seems weird

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 11]
```diff
@@ -47,7 +48,9 @@ public void stop() {}
   @Override
   protected TSStatus write(FragmentInstance fragmentInstance) {
     logger.info("Execute write plan in SchemaRegionStateMachine");
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:50]: Maybe we can remove this log

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java [hunk 12]
```diff
@@ -111,16 +114,24 @@ public QueryType getType() {
 
   public String toString() {
     StringBuilder ret = new StringBuilder();
-    ret.append(
-        String.format(
-            "FragmentInstance-%s:[Host: %s/%s]\n",
-            getId(), getHostEndpoint().getIp(), getDataRegionId().getId()));
+    ret.append(String.format("FragmentInstance-%s:", getId()));
+    if (getHostEndpoint() == null) {
+      ret.append(String.format("host endpoint has not set."));
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java:119]: Why use `String.format` as you have no parameters....

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java [hunk 13]
```diff
@@ -111,16 +114,24 @@ public QueryType getType() {
...
+    } else {
+      ret.append(String.format("host endpoint: %s.", getHostEndpoint().toString()));
+    }
+    if (getRegionReplicaSet() == null) {
+      ret.append(String.format("Region Replica set has not set.\n"));
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java:124]: same as above

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java [hunk 14]
```diff
@@ -129,30 +140,24 @@ public static FragmentInstance deserializeFrom(ByteBuffer buffer) {
...
-    fragmentInstance.hostEndpoint = endpoint;
+    fragmentInstance.regionReplicaSet = RegionReplicaSet.deserializeImpl(buffer);
+    fragmentInstance.hostEndpoint = Endpoint.deserializeImpl(buffer);
 
     return fragmentInstance;
   }
 
   @Override
   public void serializeRequest(ByteBuffer buffer) {
+    buffer.mark();
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java:151]: why use this? Is there anywhere use `buffer.reset()`?

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/PlanNodeUtil.java [hunk 15]
```diff
@@ -72,6 +72,9 @@ private static void nodeToString(PlanNode root, PrintContext ctx, StringBuilder
     }
     result.append(root.toString());
     result.append(System.lineSeparator());
+    if (root.getChildren() == null) {
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/PlanNodeUtil.java:75]: Seems better to use `for(int i = 0;i < root.getChildrenCount();i++)`

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/CreateAlignedTimeSeriesNode.java [hunk 16]
```diff
@@ -351,4 +353,24 @@ public int hashCode() {
...
+        getMeasurements(),
+        getDataTypes(),
+        getEncodings(),
+        getCompressors(),
+        getAliasList(),
+        getTagsList(),
+        getAttributesList());
+  }
+
+  //  @Override
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/CreateAlignedTimeSeriesNode.java:370]: What's this?

File: server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java [hunk 17]
```diff
@@ -43,39 +45,73 @@
...
+
+    if (fragmentInstance.getRegionReplicaSet() == null
+        || fragmentInstance.getRegionReplicaSet().isEmpty()) {
+      String msg = "Unknown regions to write, since getRegionReplicaSet is empty.";
+      LOGGER.error(msg);
+      response.setAccepted(false);
+      response.setMessage(msg);
+      return response;
+    }
+    consensusManager.addConsensusGroup(fragmentInstance.getRegionReplicaSet());
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java:88]: Do we need to call `addConsensusGroup()` each time?

File: server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java [hunk 18]
```diff
@@ -43,39 +45,73 @@
...
   }
 
   @Override
-  public TSendFragmentInstanceResp sendFragmentInstance(TSendFragmentInstanceReq req)
-      throws TException {
+  public TSendFragmentInstanceResp sendFragmentInstance(TSendFragmentInstanceReq req) {
+    TSendFragmentInstanceResp response = new TSendFragmentInstanceResp();
+    FragmentInstance fragmentInstance = null;
+    try {
+      fragmentInstance = FragmentInstance.deserializeFrom(req.fragmentInstance.body);
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/service/InternalServiceImpl.java:68]: For forwarded requests, we tend to move the deserialization logic below the consensus layer

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java (hunk 14 above)
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/FragmentInstance.java:151]: can remove. I will fix it next pr

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java (hunk 9 above)
Comment (yifuzhou) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/RequestMessage.java:64]: yes. will fix it next pr.

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/PlanNodeUtil.java (hunk 15 above)
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/PlanNodeUtil.java:75]: yes. will fix it next pr

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/CreateAlignedTimeSeriesNode.java (hunk 16 above)
Comment (yifuzhou) [reply server/src/main/java/org/apache/iotdb/db/mpp/sql/planner/plan/node/metedata/write/CreateAlignedTimeSeriesNode.java:370]: fix it in next pr

**Pull request merged**